Randomized and parallel tests

kdeldycke opened the pull request:
This PR helps detection of leaks and flaky tests by running tests in randomized order and parallel.

Results:

- Multiple cores are properly detected and tests are run in parallel:

  ```
  created: 4/4 workers
  4 workers [1336 items]
  ```

- Fix `test_echo_via_pager` test with was leaky and polluting others.

Rowlando13 commented:
Docs has the test runner as not thread safe so I am interested in what you find: 

**These tools should only be used for testing since they change the entire interpreter state for simplicity. They are not thread-safe!**
https://click.palletsprojects.com/en/stable/testing/

kdeldycke commented:
Maybe https://github.com/pallets/click/pull/1572 is fixing the issue.

pjonsson commented:
I don't think you're doing this for performance, but I investigated pytest-xdist in a different context a couple of years ago, and at the time it was difficult to get a speedup because the startup time was linear to the number of workers.

The startup time still seems to be linear to the number of workers, but the constant is much better now, here is a log with timing for `--numprocesses=`: `auto` (16), 4, 1, and a final run that still has pytest-xdist in the dependencies but without passing `--numprocesses=` to pytest.

As comparison, `uv run pytest` on stable reports `1316 passed, 21 skipped, 1 xfailed in 1.38s`  on the same machine.
```
$ uv run pytest tests/test_compat.py
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0
Using --randomly-seed=3856693423
rootdir: /home/ubuntu/src/click
configfile: pyproject.toml
plugins: xdist-3.8.0, randomly-4.0.1
16 workers [1 item]
.                                                                        [100%]
============================== 1 passed in 0.93s ===============================
$ uv run pytest tests/test_compat.py
      Built click @ file:///home/ubuntu/src/click
Uninstalled 1 package in 0.25ms
Installed 1 package in 1ms
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0
Using --randomly-seed=2447227139
rootdir: /home/ubuntu/src/click
configfile: pyproject.toml
plugins: xdist-3.8.0, randomly-4.0.1
4 workers [1 item]
.                                                                        [100%]
============================== 1 passed in 0.47s ===============================
$ uv run pytest tests/test_compat.py
      Built click @ file:///home/ubuntu/src/click
Uninstalled 1 package in 0.30ms
Installed 1 package in 1ms
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0
Using --randomly-seed=3619139103
rootdir: /home/ubuntu/src/click
configfile: pyproject.toml
plugins: xdist-3.8.0, randomly-4.0.1
1 worker [1 item]
.                                                                        [100%]
============================== 1 passed in 0.33s ===============================
$ uv run pytest tests/test_compat.py
      Built click @ file:///home/ubuntu/src/click
Uninstalled 1 package in 0.24ms
Installed 1 package in 1ms
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0
Using --randomly-seed=2375080509
rootdir: /home/ubuntu/src/click
configfile: pyproject.toml
plugins: xdist-3.8.0, randomly-4.0.1
collected 1 item

tests/test_compat.py .                                                   [100%]

============================== 1 passed in 0.02s ===============================
```

kdeldycke commented:
> The problem is that whether it uncovers something or not is completely random so if we add this to CI unrelated PR might fail CI for no apparent reason to the author.

I'm fine with that policy. I will rework this PR in a couple of days then.

kdeldycke committed:
Randomize tests and run them in parallel

Document optional test suites

kdeldycke commented:
> If you want to add a specific tox environment for this fine, but it should not run by default locally or in CI.

Ok so I just added a new `[tool.tox.env.random]` tox environment, to run this aspect of testing independently, and not in the regular CI invokation. This use the same pattern as the stress test suite I added in https://github.com/pallets/click/pull/3139 .

Now I would like to add some minimal documentation somewhere on how to run these tests. @Rowlando13 what's your opinion? Should we have a section somewhere in the Sphinx docs? I would like to document how to run the stress tests, the randomized tests, and the application smoke tests you implemented for Flask (https://github.com/pallets/click/pull/3177). So I can point contributors to them. Maybe not in `docs/testing.md` as this is for Click's users to cover how they test their own Click CLI.

davidism commented:
Adding a local `CONTRIBUTING.md` causes the standard document to be omitted. Please update the guide to place the standard text at the top first. Also, this guide is almost entirely about test environments now, it feels excessive. Something simple like the following seems easier:

> Click includes some extra test environments:
> 
> - `tox r -e stress` runs stress tests for race conditions in Click's test runner
> - `tox r -e random` runs tests in parallel in a random order to attempt to catch test pollution.

kdeldycke commented:
> Adding a local `CONTRIBUTING.md` causes the standard document to be omitted. Please update the guide to place the standard text at the top first. Also, this guide is almost entirely about test environments now, it feels excessive. Something simple like the following seems easier:

OK I see. To avoid confusion and spread the contribution content everywhere, I regrouped everything in the `contributing.md` file and simplified it.

davidism commented:
This is the file I'm referring to. https://github.com/pallets/.github/blob/main/CONTRIBUTING.md GitHub automatically links to this file if the file is not present in a project. At least the introduction section needs to be preserved to explain what this contributing guide is and is not and then you also need to link to the copy of the quick guide on the website. I want everyone to be getting as standard experience across all projects as possible.

kdeldycke commented:
> This is the file I'm referring to. https://github.com/pallets/.github/blob/main/CONTRIBUTING.md GitHub automatically links to this file if the file is not present in a project. At least the introduction section needs to be preserved to explain what this contributing guide is and is not and then you also need to link to the copy of the quick guide on the website. I want everyone to be getting as standard experience across all projects as possible.

I'm not 100% certain about what you're asking for. I get that the `Contributing` tab on `click` GitHub page is sourced from `https://github.com/pallets/.github/blob/main/CONTRIBUTING.md`:

<img width="1092" height="331" alt="Screenshot 2026-04-09 at 16 36 53" src="https://github.com/user-attachments/assets/51c4be5c-63c3-44d3-8392-ed2049e92b14" />

But what about the Click-specific changes? I guess you don't want me to put these specific stuff there and contribute to the`pallets/.github` repository? So in the mean time I just altered the `docs/contributing.md` introduction to remind readers that this page contains Click-specific stuff only.

davidism commented:
Sorry about that, I was on my phone and I think I missed that you were editing `docs/contributing.md` instead of `CONTRIBUTING.md`. In that case, you can absolutely put Click specific stuff in the docs, although keeping that intro paragraph is a good idea too.

kdeldycke commented:
> Sorry about that, I was on my phone and I think I missed that you were editing `docs/contributing.md` instead of `CONTRIBUTING.md`. In that case, you can absolutely put Click specific stuff in the docs, although keeping that intro paragraph is a good idea too.

Good! And yes, it is not a bad idea to remind users that global contributing guidelines exists for all Pallets projects. Also a good reminder for maintainers like me to check the global guidelines before adding content to `docs/contributing.md`! 😉

So with all comments addressed, this PR is ready to be merged upstream! :)

kdeldycke commented:
Thanks @Rowlando13 for the merge! :)
